File: tests/src/main/javascript/util/MockMslContext.js
/**
 * Copyright (c) 2012-2017 Netflix, Inc.  All rights reserved.
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * MSL context for unit tests.
 * 
 * @author Wesley Miaw <wmiaw@netflix.com>
 */
(function(require, module) {
    "use strict";
    
    var MslContext = require('msl-core/util/MslContext.js');
    var AsyncExecutor = require('msl-core/util/AsyncExecutor.js');
    var EntityAuthenticationScheme = require('msl-core/entityauth/EntityAuthenticationScheme.js');
    var UserAuthenticationScheme = require('msl-core/userauth/UserAuthenticationScheme.js');
    var KeyExchangeScheme = require('msl-core/keyx/KeyExchangeScheme.js');
    
	/** MSL encryption key. */
    var MSL_ENCRYPTION_KEY = "HVjzuPdH0Wqxk8TApiTqzw==";
    /** MSL HMAC key. */
    var MSL_HMAC_KEY = "166/1YebsOCtAWpM88s5gvW6Jg2lICRbtCJ1vXlHNww=";
    /** MSL wrapping key. */
    var MSL_WRAPPING_KEY = "g7aaFYDTI6LnndmyJiaz9g==";
	
	var MockMslContext = module.exports = MslContext.extend({
		/**
	     * Create a new test MSL context.
	     * 
	     * @param {EntityAuthenticationScheme} scheme entity authentication scheme.
	     * @param {boolean} peerToPeer true if the context should operate in peer-
	     *        to-peer mode.
	     * @param {result: function(MockMslContext), error: function(Error)}
	     *        callback the callback that will receive the mock MSL context or
	     *        any thrown exceptions.
	     * @throws CryptoException if the MSL crypto keys cannot be created.
	     * @throws MslCryptoException if there is an error signing or creating the
	     *         entity authentication data.
	     * @throws MslEncodingException if there is an error creating the entity
	     *         authentication data.
	     */
		init: function init(scheme, peerToPeer, callback) {
		    var UnauthenticatedAuthenticationFactory = require('msl-core/entityauth/UnauthenticatedAuthenticationFactory.js');
		    var UnauthenticatedSuffixedAuthenticationFactory = require('msl-core/entityauth/UnauthenticatedSuffixedAuthenticationFactory.js');
		    var MasterTokenProtectedAuthenticationFactory = require('msl-core/entityauth/MasterTokenProtectedAuthenticationFactory.js');
		    var ProvisionedAuthenticationFactory = require('msl-core/entityauth/ProvisionedAuthenticationFactory.js');
		    var SecretKey = require('msl-core/crypto/SecretKey.js');
		    var WebCryptoAlgorithm = require('msl-core/crypto/WebCryptoAlgorithm.js');
		    var WebCryptoUsage = require('msl-core/crypto/WebCryptoUsage.js');
		    var MslInternalException = require('msl-core/MslInternalException.js');
		    var PresharedAuthenticationData = require('msl-core/entityauth/PresharedAuthenticationData.js');
		    var PresharedProfileAuthenticationData = require('msl-core/entityauth/PresharedProfileAuthenticationData.js');
		    var X509AuthenticationData = require('msl-core/entityauth/X509AuthenticationData.js');
		    var RsaAuthenticationData = require('msl-core/entityauth/RsaAuthenticationData.js');
		    var EccAuthenticationData = require('msl-core/entityauth/EccAuthenticationData.js');
		    var UnauthenticatedAuthenticationData = require('msl-core/entityauth/UnauthenticatedAuthenticationData.js');
		    var UnauthenticatedSuffixedAuthenticationData = require('msl-core/entityauth/UnauthenticatedSuffixedAuthenticationData.js');
		    var MessageCapabilities = require('msl-core/msg/MessageCapabilities.js');
		    var MslConstants = require('msl-core/MslConstants.js');
		    var MslEncoderFormat = require('msl-core/io/MslEncoderFormat.js');
		    var SymmetricCryptoContext = require('msl-core/crypto/SymmetricCryptoContext.js');
		    var AsymmetricWrappedExchange = require('msl-core/keyx/AsymmetricWrappedExchange.js');
		    var SymmetricWrappedExchange = require('msl-core/keyx/SymmetricWrappedExchange.js');
		    var DiffieHellmanExchange = require('msl-core/keyx/DiffieHellmanExchange.js');
		    var SimpleMslStore = require('msl-core/util/SimpleMslStore.js');
		    var DefaultMslEncoderFactory = require('msl-core/io/DefaultMslEncoderFactory.js');
		    var Random = require('msl-core/util/Random.js');
		    
		    var MockAuthenticationUtils = require('../util/MockAuthenticationUtils.js');
            var MockIdentityProvisioningService = require('../entityauth/MockIdentityProvisioningService.js');
		    var MockPresharedAuthenticationFactory = require('../entityauth/MockPresharedAuthenticationFactory.js');
		    var MockPresharedProfileAuthenticationFactory = require('../entityauth/MockPresharedProfileAuthenticationFactory.js');
		    var MockRsaAuthenticationFactory = require('../entityauth/MockRsaAuthenticationFactory.js');
		    var MockEccAuthenticationFactory = require('../entityauth/MockEccAuthenticationFactory.js');
		    var MockX509AuthenticationFactory = require('../entityauth/MockX509AuthenticationFactory.js');
		    var MockUnauthenticatedAuthenticationFactory = require('../entityauth/MockUnauthenticatedAuthenticationFactory.js');
		    var MockEmailPasswordAuthenticationFactory = require('../userauth/MockEmailPasswordAuthenticationFactory.js');
		    var MockUserIdTokenAuthenticationFactory = require('../userauth/MockUserIdTokenAuthenticationFactory.js');
		    var MockTokenFactory = require('../tokens/MockTokenFactory.js');
		    var MockDiffieHellmanParameters = require('../keyx/MockDiffieHellmanParameters.js');
		    
		    var self = this;

		    // Set up entity authentication factories.
		    AsyncExecutor(callback, function pskAuthFactory() {
                var authutils = new MockAuthenticationUtils();
		        var entityAuthFactories = {};
		        MockPresharedAuthenticationFactory.create({
		            result: function(factory) {
		                AsyncExecutor(callback, function() {
		                    entityAuthFactories[EntityAuthenticationScheme.PSK.name] = factory;
		                    pskProfileAuthFactory(authutils, entityAuthFactories);
		                }, self);
		            },
		            error: callback.error,
		        });
		    });
		    function pskProfileAuthFactory(authutils, entityAuthFactories) {
		        MockPresharedProfileAuthenticationFactory.create({
                    result: function(factory) {
                        AsyncExecutor(callback, function() {
                            entityAuthFactories[EntityAuthenticationScheme.PSK_PROFILE.name] = factory;
                            rsaAuthFactory(authutils, entityAuthFactories);
                        }, self);
                    },
                    error: callback.error,
                });
		    }
		    function rsaAuthFactory(authutils, entityAuthFactories) {
		        MockRsaAuthenticationFactory.create(null, {
		            result: function(factory) {
		                AsyncExecutor(callback, function() {
		                    entityAuthFactories[EntityAuthenticationScheme.RSA.name] = factory;
		                    eccAuthFactory(authutils, entityAuthFactories);
		                }, self);
		            },
		            error: callback.error,
		        });
		    }
		    function eccAuthFactory(authutils, entityAuthFactories) {
                MockEccAuthenticationFactory.create(null, {
                    result: function(factory) {
                        AsyncExecutor(callback, function() {
                            entityAuthFactories[EntityAuthenticationScheme.ECC.name] = factory;
                            syncEntityAuthFactories(authutils, entityAuthFactories);
                        }, self);
                    },
                    error: callback.error,
                });
		    }
		    function syncEntityAuthFactories(authutils, entityAuthFactories) {
		        entityAuthFactories[EntityAuthenticationScheme.X509.name] = new MockX509AuthenticationFactory();
		        entityAuthFactories[EntityAuthenticationScheme.NONE.name] = new MockUnauthenticatedAuthenticationFactory();
		        entityAuthFactories[EntityAuthenticationScheme.NONE_SUFFIXED.name] = new UnauthenticatedSuffixedAuthenticationFactory();
		        entityAuthFactories[EntityAuthenticationScheme.MT_PROTECTED.name] = new MasterTokenProtectedAuthenticationFactory(authutils);
		        entityAuthFactories[EntityAuthenticationScheme.PROVISIONED.name] = new ProvisionedAuthenticationFactory(new MockIdentityProvisioningService(this));
                syncUserAuthFactories(authutils, entityAuthFactories);
		    }
		    function syncUserAuthFactories(authutils, entityAuthFactories) {
		        var userAuthFactories = {};
		        userAuthFactories[UserAuthenticationScheme.EMAIL_PASSWORD.name] = new MockEmailPasswordAuthenticationFactory();
		        userAuthFactories[UserAuthenticationScheme.USER_ID_TOKEN.name] = new MockUserIdTokenAuthenticationFactory();
                mslCryptoContext(authutils, entityAuthFactories, userAuthFactories);
		    }
		    function mslCryptoContext(authutils, entityAuthFactories, userAuthFactories) {
		        SecretKey.import(MSL_ENCRYPTION_KEY, WebCryptoAlgorithm.AES_CBC, WebCryptoUsage.ENCRYPT_DECRYPT, {
		            result: function (mslEncryptionKey) {
		                SecretKey.import(MSL_HMAC_KEY, WebCryptoAlgorithm.HMAC_SHA256, WebCryptoUsage.SIGN_VERIFY, {
		                    result: function (mslHmacKey) {
		                        SecretKey.import(MSL_WRAPPING_KEY, WebCryptoAlgorithm.A128KW, WebCryptoUsage.WRAP_UNWRAP, {
		                            result: function(mslWrappingKey) {
		                                finish(authutils, entityAuthFactories, userAuthFactories, mslEncryptionKey, mslHmacKey, mslWrappingKey);
		                            },
		                            error: function(e) {
		                                callback.error(new MslInternalException("Unable to create MSL wrap key.", e));
		                            }
		                        });
		                    },
		                    error: function (e) {
		                        callback.error(new MslInternalException("Unable to create MSL HMAC key.", e));
		                    }
		                });
		            },
		            error: function (e) {
		                callback.error(new MslInternalException("Unable to create MSL encryption key.", e));
		            }
		        });
		    }
		    function finish(authutils, entityAuthFactories, userAuthFactories, mslEncryptionKey, mslHmacKey, mslWrapKey) {
		        AsyncExecutor(callback, function() {
		            // Set up entity authentication data.
		            var entityAuthData;
		            if (EntityAuthenticationScheme.PSK == scheme) {
		                entityAuthData = new PresharedAuthenticationData(MockPresharedAuthenticationFactory.PSK_ESN);
		            } else if (EntityAuthenticationScheme.PSK_PROFILE == scheme) {
		                entityAuthData = new PresharedProfileAuthenticationData(MockPresharedProfileAuthenticationFactory.PSK_ESN, MockPresharedProfileAuthenticationFactory.PROFILE);
		            } else if (EntityAuthenticationScheme.X509 == scheme) {
		                entityAuthData = new X509AuthenticationData(MockX509AuthenticationFactory.X509_CERT);
		            } else if (EntityAuthenticationScheme.RSA == scheme) {
		                entityAuthData = new RsaAuthenticationData(MockRsaAuthenticationFactory.RSA_ESN, MockRsaAuthenticationFactory.RSA_PUBKEY_ID);
		            } else if (EntityAuthenticationScheme.ECC == scheme) {
		                entityAuthData = new EccAuthenticationData(MockRsaAuthenticationFactory.ECC_ESN, MockRsaAuthenticationFactory.ECC_PUBKEY_ID);
		            } else if (EntityAuthenticationScheme.NONE == scheme) {
		                entityAuthData = new UnauthenticatedAuthenticationData("MOCKUNAUTH-ESN");
		            } else if (EntityAuthenticationScheme.NONE_SUFFIXED == scheme) {
		                entityAuthData = new UnauthenticatedSuffixedAuthenticationData("MOCKUNAUTH-ROOT", "MOCKUNAUTH-SUFFIX");
		            } else {
		                throw new MslInternalException("Unsupported authentication type: " + scheme);
		            }

		            // Set message capabilities.
		            var capabilities = new MessageCapabilities([MslConstants.CompressionAlgorithm.LZW], [ "en-US" ], [MslEncoderFormat.JSON]);

		            // Set the MSL crypto context.
		            var mslCryptoContext = new SymmetricCryptoContext(this, "TestMslKeys", mslEncryptionKey, mslHmacKey, mslWrapKey);

		            // Set up token factory.
		            var tokenFactory = new MockTokenFactory();

		            // Set up Diffie-Hellman parameter specifications.
		            var paramSpecs = MockDiffieHellmanParameters.getDefaultParameters();

		            // Set up key exchange factories.
		            var keyxFactories = [];
		            keyxFactories.push(new AsymmetricWrappedExchange(authutils));
		            keyxFactories.push(new SymmetricWrappedExchange(authutils));
		            keyxFactories.push(new DiffieHellmanExchange(paramSpecs, authutils));

		            // Set up the MSL store.
		            var store = new SimpleMslStore();
		            
		            // Set up the MSL encoder factory.
		            var encoderFactory = new DefaultMslEncoderFactory();

		            // The properties.
		            var props = {
		                _random: { value: new Random(), writable: false, enumerable: false, configurable: false },
		                _mslCryptoContext: { value: mslCryptoContext, writable: true, enumerable: false, configurable: false },
		                _peerToPeer: { value: peerToPeer, writable: false, enumerable: false, configurable: false },
		                _capabilities: { value: capabilities, writable: true, enumerable: false, configurable: false },
		                _entityAuthData: { value: entityAuthData, writable: true, enumerable: false, configurable: false },
		                _entityAuthFactories: { value: entityAuthFactories, writable: false, enumerable: false, configurable: false },
		                _userAuthFactories: { value: userAuthFactories, writable: false, enumerable: false, configurable: false },
		                _tokenFactory: { value: tokenFactory, writable: true, enumerable: false, configurable: false },
		                _paramSpecs: { value: paramSpecs, writable: false, enumerable: false, configurable: false },
		                _keyxFactories: { value: keyxFactories, writable: false, enumerable: false, configurable: false },
		                _store: { value: store, writable: true, enumerable: false, configurable: false },
		                _encoderFactory: { value: encoderFactory, writable: true, enumerable: false, configurable: false },
		            };
		            Object.defineProperties(this, props);

		            // Return the mock MSL context.
		            return this;
		        }, self);
		    }
		},
		
		/** @inheritDoc */
		getTime: function getTime() {
			return Date.now();
		},

		/** @inheritDoc */
		getRandom: function getRandom() {
			return this._random;
		},

		/** @inheritDoc */
		isPeerToPeer: function isPeerToPeer() {
			return this._peerToPeer;
		},
		
	    /**
	     * Set the message capabilities.
	     * 
	     * @param {MessageCapabilities} capabilities the new message capabilities.
	     */
	    setMessageCapabilities: function setMessageCapabilities(capabilities) {
	        this._capabilities = capabilities;
	    },
		
		/** @inheritDoc */
		getMessageCapabilities: function getMessageCapabilities() {
			return this._capabilities;
		},

		setEntityAuthenticationData: function setEntityAuthenticationData(entityAuthData) {
			this._entityAuthData = entityAuthData;
		},
		
		/** @inheritDoc */
		getEntityAuthenticationData: function getEntityAuthenticationData(reauthCode, callback) {
			callback.result(this._entityAuthData);
		},
		
	    /**
	     * Set the MSL crypto context.
	     * 
	     * @param {ICryptoContext} cryptoContext the new MSL crypto context.
	     */
	    setMslCryptoContext: function setMslCryptoContext(cryptoContext) {
	        this._mslCryptoContext = cryptoContext;
	    },

		/** @inheritDoc */
		getMslCryptoContext: function getMslCryptoContext() {
			return this._mslCryptoContext;
		},
        
        /** @inheritDoc */
		getEntityAuthenticationScheme: function getEntityAuthenticationScheme(name) {
		    return EntityAuthenticationScheme.getScheme(name);
		},

		/**
		 * Adds or replaces the entity authentication factory associated with the
		 * entity authentication scheme of the provided factory.
		 * 
		 * @param {EntityAuthenticationFactory} factory entity authentication factory.
		 */
		addEntityAuthenticationFactory: function addEntityAuthenticationFactory(factory) {
			this._entityAuthFactories[factory.scheme.name] = factory;
		},

		/**
		 * Removes the entity authentication factory associated with the specified
		 * entity authentication scheme.
		 * 
		 * @param {EntityAuthenticationScheme} scheme entity authentication scheme.
		 */
		removeEntityAuthenticationFactory: function removeEntityAuthenticationFactory(scheme) {
			delete this._entityAuthFactories[scheme.name];
		},

		/** @inheritDoc */
		getEntityAuthenticationFactory: function getEntityAuthenticationFactory(scheme) {
			return this._entityAuthFactories[scheme.name];
		},
		
		/** @inheritDoc */
		getUserAuthenticationScheme: function getUserAuthenticationScheme(name) {
		    return UserAuthenticationScheme.getScheme(name);
		},

		/**
		 * Adds or replaces the user authentication factory associated with the
		 * user authentication scheme of the provided factory.
		 * 
		 * @param {UserAuthenticationFactory} factory user authentication factory.
		 */
		addUserAuthenticationFactory: function addUserAuthenticationFactory(factory) {
			this._userAuthFactories[factory.scheme.name] = factory;
		},

		/**
		 * Removes the user authentication factory associated with the specified
		 * user authentication scheme.
		 * 
		 * @param {UserAuthenticationScheme} scheme user authentication scheme.
		 */
		removeUserAuthenticationFactory: function removeUserAuthenticationFactory(scheme) {
			delete this._userAuthFactories[scheme.name];
		},

		/** @inheritDoc */
		getUserAuthenticationFactory: function getUserAuthenticationFactory(scheme) {
			return this._userAuthFactories[scheme.name];
		},
	    
	    /**
	     * Sets the token factory.
	     * 
	     * @param {TokenFactory} factory the token factory.
	     */
	    setTokenFactory: function setTokenFactory(factory) {
	        this._tokenFactory = factory;
	    },

		/** @inheritDoc */
		getTokenFactory: function getTokenFactory() {
			return this._tokenFactory;
		},
		
		/**
		 * @return {Object.<string,DhParameterSpec>} the Diffie-Hellman
		 *         parameter specifications supported by the Diffie-Hellman
		 *         key exchange factory.
		 */
		getDhParameterSpecs: function getDhParameterSpecs() {
			return this._paramSpecs;
		},
		
		/** @inheritDoc */
		getKeyExchangeScheme: function getKeyExchangeScheme(name) {
		    return KeyExchangeScheme.getScheme(name);
		},

        /**
         * Adds a key exchange factory to the end of the preferred set.
         * 
         * @param {KeyExchangeFactory} factory key exchange factory.
         */
        addKeyExchangeFactory: function addKeyExchangeFactory(factory) {
            this._keyxFactories.push(factory);
        },

        /**
         * Removes all key exchange factories associated with the specified key
         * exchange scheme.
         * 
         * @param {KeyExchangeScheme} scheme key exchange scheme.
         */
        removeKeyExchangeFactories: function removeKeyExchangeFactories(scheme) {
            for (var i = 0; i < this._keyxFactories.length; ++i) {
                var factory = this._keyxFactories[i];
                if (factory.scheme == scheme) {
                    this._keyxFactories.splice(i, 1);
                    --i;
                }
            }
        },

		/** @inheritDoc */
		getKeyExchangeFactory: function getKeyExchangeFactory(scheme) {
			for (var i = 0; i < this._keyxFactories.length; ++i) {
				var factory = this._keyxFactories[i];
				if (factory.scheme == scheme)
					return factory;
			}
			return undefined;
		},

		/** @inheritDoc */
		getKeyExchangeFactories: function getKeyExchangeFactories(scheme) {
			return this._keyxFactories;
		},
		
		/**
		 * Sets the MSL store.
		 * 
		 * @param {MslStore} store the MSL store.
		 */
		setMslStore: function setMslStore(store) {
		    this._store = store;
		},

		/** @inheritDoc */
		getMslStore: function getMslStore() {
			return this._store;
		},
		
		/**
		 * Sets the MSL encoder factory.
		 * 
		 * @param {MslEncoderFactory} encoderFactory the MSL encoder factory.
		 */
		setMslEncoderFactory: function setMslEncoderFactory(encoderFactory) {
		    this._encoderFactory = encoderFactory;
		},
		
		/** @inheritDoc */
		getMslEncoderFactory: function getMslEncoderFactory() {
		    return this._encoderFactory;
		},
	});
	
    /**
     * Create a new test MSL context.
     * 
     * @param {EntityAuthenticationScheme} scheme entity authentication scheme.
     * @param {boolean} peerToPeer true if the context should operate in peer-
     *        to-peer mode.
     * @param {result: function(MockMslContext), error: function(Error)}
     *        callback the callback that will receive the mock MSL context or
     *        any thrown exceptions.
     * @throws CryptoException if the MSL crypto keys cannot be created.
     * @throws MslCryptoException if there is an error signing or creating the
     *         entity authentication data.
     * @throws MslEncodingException if there is an error creating the entity
     *         authentication data.
     */
	var MockMslContext$create = function MockMslContext$create(scheme, peerToPeer, callback) {
	    new MockMslContext(scheme, peerToPeer, callback);
	};
	
	// Exports.
	module.exports.create = MockMslContext$create;
})(require, (typeof module !== 'undefined') ? module : mkmodule('MockMslContext'));

File: tests/src/main/cpp/util/MockMslContext.cpp
/**
 * Copyright (c) 2016-2017 Netflix, Inc.  All rights reserved.
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "MockMslContext.h"
#include <crypto/Key.h>
#include <crypto/JcaAlgorithm.h>
#include <crypto/Random.h>
#include <crypto/SymmetricCryptoContext.h>
#include <entityauth/EntityAuthenticationScheme.h>
#include <entityauth/PresharedAuthenticationData.h>
#include <entityauth/PresharedProfileAuthenticationData.h>
#include <entityauth/RsaAuthenticationData.h>
#include <entityauth/UnauthenticatedAuthenticationData.h>
#include <entityauth/UnauthenticatedAuthenticationFactory.h>
#include <io/DefaultMslEncoderFactory.h>
#include <keyx/AsymmetricWrappedExchange.h>
#include <keyx/DiffieHellmanExchange.h>
#include <keyx/KeyExchangeScheme.h>
#include <keyx/KeyExchangeFactory.h>
#include <keyx/SymmetricWrappedExchange.h>
#include <msg/MessageCapabilities.h>
#include <MslInternalException.h>
#include <userauth/UserAuthenticationScheme.h>
#include <userauth/UserAuthenticationFactory.h>
#include <stdint.h>
#include <sys/time.h>
#include <util/SimpleMslStore.h>
#include <map>
#include <memory>
#include <set>
#include <string>
#include <vector>

#include "../tokens/MockTokenFactory.h"
#include "../entityauth/MockPresharedAuthenticationFactory.h"
#include "../entityauth/MockPresharedProfileAuthenticationFactory.h"
#include "../entityauth/MockRsaAuthenticationFactory.h"
#include "../keyx/MockDiffieHellmanParameters.h"
#include "../userauth/MockEmailPasswordAuthenticationFactory.h"
#include "../userauth/MockUserIdTokenAuthenticationFactory.h"
#include "../util/MockAuthenticationUtils.h"

using netflix::msl::msg::MessageCapabilities;

using namespace std;
using namespace testing;
using namespace netflix::msl;
using namespace netflix::msl::crypto;
using namespace netflix::msl::entityauth;
using namespace netflix::msl::io;
using namespace netflix::msl::keyx;
using namespace netflix::msl::msg;
using namespace netflix::msl::tokens;
using namespace netflix::msl::userauth;
using namespace netflix::msl::util;

namespace netflix {
namespace msl {
namespace util {

namespace {

/** MSL encryption key. */
const uint8_t MSL_ENCRYPTION_KEY[] = {
    0x1d, 0x58, 0xf3, 0xb8, 0xf7, 0x47, 0xd1, 0x6a,
    0xb1, 0x93, 0xc4, 0xc0, 0xa6, 0x24, 0xea, 0xcf,
};
/** MSL HMAC key. */
const uint8_t MSL_HMAC_KEY[] = {
    0xd7, 0xae, 0xbf, 0xd5, 0x87, 0x9b, 0xb0, 0xe0,
    0xad, 0x01, 0x6a, 0x4c, 0xf3, 0xcb, 0x39, 0x82,
    0xf5, 0xba, 0x26, 0x0d, 0xa5, 0x20, 0x24, 0x5b,
    0xb4, 0x22, 0x75, 0xbd, 0x79, 0x47, 0x37, 0x0c,
};
/** MSL wrapping key. */
const uint8_t MSL_WRAPPING_KEY[] = {
    0x83, 0xb6, 0x9a, 0x15, 0x80, 0xd3, 0x23, 0xa2,
    0xe7, 0x9d, 0xd9, 0xb2, 0x26, 0x26, 0xb3, 0xf6,
};

template<typename T, size_t S> size_t SizeOf(T(&)[S]) { return S; }
#define BYTE_ARRAY(a) make_shared<ByteArray>(a, a + SizeOf(a))

} // namespace anonymous


MockMslContext::MockMslContext(const EntityAuthenticationScheme& scheme, bool peerToPeer)
: random(make_shared<Random>())
, peerToPeer(peerToPeer)
{
    if (scheme == EntityAuthenticationScheme::PSK)
        entityAuthData = make_shared<PresharedAuthenticationData>(MockPresharedAuthenticationFactory::PSK_ESN);
    else if (scheme == EntityAuthenticationScheme::PSK_PROFILE)
        entityAuthData = make_shared<PresharedProfileAuthenticationData>(MockPresharedProfileAuthenticationFactory::PSK_ESN, MockPresharedProfileAuthenticationFactory::PROFILE);
//    else if (scheme == EntityAuthenticationScheme::X509)
//        entityAuthData = make_shared<X509AuthenticationData>(MockX509AuthenticationFactory::X509_CERT);
    else if (scheme == EntityAuthenticationScheme::RSA)
        entityAuthData = make_shared<RsaAuthenticationData>(MockRsaAuthenticationFactory::RSA_ESN, MockRsaAuthenticationFactory::RSA_PUBKEY_ID);
    else if (scheme == EntityAuthenticationScheme::NONE)
        entityAuthData = make_shared<UnauthenticatedAuthenticationData>("MOCKUNAUTH-ESN");
//    else if (scheme == EntityAuthenticationScheme::NONE_SUFFIXED)
//        entityAuthData = make_shared<UnauthenticatedSuffixedAuthenticationData>("MOCKUNAUTH-ROOT", "MOCKUNAUTH-SUFFIX");

    set<MslConstants::CompressionAlgorithm> algos;
    algos.insert(MslConstants::CompressionAlgorithm::GZIP);
    algos.insert(MslConstants::CompressionAlgorithm::LZW);
    vector<string> languages;
    languages.insert(languages.begin(), "en-US");
    set<MslEncoderFormat> formats;
    formats.insert(MslEncoderFormat::JSON);
    capabilities = make_shared<MessageCapabilities>(algos, languages, formats);

    tokenFactory = make_shared<MockTokenFactory>();
    store = make_shared<SimpleMslStore>();
    encoderFactory = make_shared<DefaultMslEncoderFactory>();

    shared_ptr<DiffieHellmanParameters> params = MockDiffieHellmanParameters::getDefaultParameters();
    authutils = make_shared<MockAuthenticationUtils>();

    entityAuthFactories.insert(make_pair(EntityAuthenticationScheme::PSK, make_shared<MockPresharedAuthenticationFactory>()));
    entityAuthFactories.insert(make_pair(EntityAuthenticationScheme::PSK_PROFILE, make_shared<MockPresharedProfileAuthenticationFactory>()));
    entityAuthFactories.insert(make_pair(EntityAuthenticationScheme::RSA, make_shared<MockRsaAuthenticationFactory>()));
    entityAuthFactories.insert(make_pair(EntityAuthenticationScheme::NONE, make_shared<UnauthenticatedAuthenticationFactory>(authutils)));
//    entityAuthFactories.insert(make_pair(EntityAuthenticationScheme::X509, make_shared<MockX509AuthenticationFactory>()));
//    entityAuthFactories.insert(make_pair(EntityAuthenticationScheme::NONE_SUFFIXED, make_shared<UnauthenticatedSuffixedAuthenticationFactory>(authutils)));
//    entityAuthFactories.insert(make_pair(EntityAuthenticationScheme::MT_PROTECTED, make_shared<MasterTokenProtectedAuthenticationFactory>(authutils)));
//    entityAuthFactories.insert(make_pair(EntityAuthenticationScheme::PROVISIONED, make_shared<ProvisionedAuthenticationFactory>(new MockIdentityProvisioningService(this))));

    userAuthFactories.insert(make_pair(UserAuthenticationScheme::EMAIL_PASSWORD, make_shared<MockEmailPasswordAuthenticationFactory>()));
    userAuthFactories.insert(make_pair(UserAuthenticationScheme::USER_ID_TOKEN, make_shared<MockUserIdTokenAuthenticationFactory>()));

    keyxFactories.insert(make_shared<AsymmetricWrappedExchange>(authutils));
    keyxFactories.insert(make_shared<SymmetricWrappedExchange>(authutils));
    keyxFactories.insert(make_shared<DiffieHellmanExchange>(params, authutils));
}

int64_t MockMslContext::getTime()
{
    struct timeval tp;
    gettimeofday(&tp, NULL);
    uint64_t ms = static_cast<uint64_t>(tp.tv_sec) * 1000ull + static_cast<uint64_t>(tp.tv_usec) / 1000ull;
    return static_cast<int64_t>(ms);
}

shared_ptr<ICryptoContext> MockMslContext::getMslCryptoContext()
{
	if (!mslCryptoContext) {
	    const SecretKey mslEncryptionKey(BYTE_ARRAY(MSL_ENCRYPTION_KEY), JcaAlgorithm::AES);
	    const SecretKey mslHmacKey(BYTE_ARRAY(MSL_HMAC_KEY), JcaAlgorithm::HMAC_SHA256);
	    const SecretKey mslWrappingKey(BYTE_ARRAY(MSL_WRAPPING_KEY), JcaAlgorithm::AESKW);
	    mslCryptoContext = make_shared<SymmetricCryptoContext>(shared_from_this(), "TestMslKeys", mslEncryptionKey, mslHmacKey, mslWrappingKey);
	}
	return mslCryptoContext;
}

EntityAuthenticationScheme MockMslContext::getEntityAuthenticationScheme(const string& name)
{
    return EntityAuthenticationScheme::getScheme(name);
}

void MockMslContext::addEntityAuthenticationFactory(shared_ptr<EntityAuthenticationFactory> factory)
{
	const EntityAuthenticationScheme scheme = factory->getScheme();
	entityAuthFactories.erase(scheme);
    entityAuthFactories.insert(make_pair(scheme, factory));
}

void MockMslContext::removeEntityAuthenticationFactory(const EntityAuthenticationScheme& scheme)
{
    entityAuthFactories.erase(scheme);
}

shared_ptr<EntityAuthenticationFactory> MockMslContext::getEntityAuthenticationFactory(const EntityAuthenticationScheme& scheme)
{
    map<EntityAuthenticationScheme, shared_ptr<EntityAuthenticationFactory>>::const_iterator it = entityAuthFactories.find(scheme);
    return (it != entityAuthFactories.end()) ? it->second : shared_ptr<EntityAuthenticationFactory>();
}

UserAuthenticationScheme MockMslContext::getUserAuthenticationScheme(const string& name)
{
    return UserAuthenticationScheme::getScheme(name);
}

void MockMslContext::addUserAuthenticationFactory(shared_ptr<UserAuthenticationFactory> factory)
{
	const UserAuthenticationScheme scheme = factory->getScheme();
	userAuthFactories.erase(scheme);
    userAuthFactories.insert(make_pair(scheme, factory));
}

void MockMslContext::removeUserAuthenticationFactory(const UserAuthenticationScheme& scheme)
{
    userAuthFactories.erase(scheme);
}

shared_ptr<UserAuthenticationFactory> MockMslContext::getUserAuthenticationFactory(const UserAuthenticationScheme& scheme)
{
    map<UserAuthenticationScheme, shared_ptr<UserAuthenticationFactory>>::const_iterator it = userAuthFactories.find(scheme);
    return (it != userAuthFactories.end()) ? it->second : shared_ptr<UserAuthenticationFactory>();
}

KeyExchangeScheme MockMslContext::getKeyExchangeScheme(const std::string& name)
{
    return KeyExchangeScheme::getScheme(name);
}

void MockMslContext::addKeyExchangeFactory(shared_ptr<KeyExchangeFactory> factory)
{
    pair<set<shared_ptr<keyx::KeyExchangeFactory>>::iterator, bool> result =
            keyxFactories.insert(factory);
    if (!result.second)
        throw MslInternalException("Could not add KeyExchangeFactory");
}

void MockMslContext::removeKeyExchangeFactories(const KeyExchangeScheme& scheme)
{
    // FIXME: use std::remove_if algorithm?
    for (set<shared_ptr<KeyExchangeFactory>>::iterator it = keyxFactories.begin(); it != keyxFactories.end(); )
    {
        if ((*it)->getScheme() == scheme) {
            keyxFactories.erase(it++);
        }
        else {
            ++it;
        }
    }
}

shared_ptr<KeyExchangeFactory> MockMslContext::getKeyExchangeFactory(const KeyExchangeScheme& scheme)
{
    for (set<shared_ptr<KeyExchangeFactory>>::const_iterator it = keyxFactories.begin(); it != keyxFactories.end(); ++it)
    {
        if ((*it)->getScheme() == scheme)
            return *it;
    }
    return shared_ptr<KeyExchangeFactory>();
}

set<shared_ptr<KeyExchangeFactory>> MockMslContext::getKeyExchangeFactories()
{
    return keyxFactories;
}

}}} // namespace netflix::msl::util
